doc: tweak context reference doc (#13218)
ocaml/dune: doc/reference/dune-workspace/context.rst

Reviewer: is a path interpreted relative to (no hyphen)

context
-------

The ``(context ...)`` stanza declares a build context. The kind of context can
be either ``(default ...)`` for the regular build context, or ``(opam
...)`` to use an opam switch.

The ``default`` context without extra fields can be specifed as ``(context
default)``

.. code:: dune

    (context
      (default
        <optional-fields>))

.. code:: dune

    (context
      (opam
        (switch <opam-switch-name>)
          <optional-fields>))

``<optional-fields>`` common to both context types are:

-  ``(name <name>)`` is the subdirectory's name for ``_build``, where this

>    build's context artifacts will be stored. If omitted, the context name

Reviewer: So   
- For (context default): If omitted, the name is default
- For (context (opam ...)): If omitted, the name is derived from the switch name

   ``default`` for the default context, and switch name for the opam context.

- ``(merlin)`` instructs Dune to use this build context for Merlin.

- ``(generate_merlin_rules)`` instructs Dune to generate Merlin rules for this
  context, even if it is not the one selected via ``(merlin)``.

- ``(profile <profile>)`` sets a different profile for a :term:`build context`.

>   This has precedence over the command-line option ``--profile``.

Reviewer 2: This seems like a design error. What is the rationale for this?

>   This has precedence over the command-line option ``--profile``.

Reviewer 2: Thanks. I think this is out of scope for @Sudha247's PR, but IMO this is quite counter-intuitive, both based on all other tools I'm familiar with and based on principle. We should probably add a followup that motivates this precedence for users. As a further followup, we should consider whether this is actually a well reasoned, usable design decision. This will relate to https://github.com/ocaml/dune/issues/12819

context

> -------

Reviewer 3: I think it would be nice to use @Alizter's description on what a context is here.

context
-------

The ``(context ...)`` stanza allows you to declare multiple build contexts which
can be built at the same time. By default Dune uses a single context called

> ``default`` whose build directory appears in ``_build/default/``. There are two

Reviewer: ```suggestion
``default`` whose build artefacts appear in ``_build/default/``. There are two
```

build context, or ``(opam ...)`` to use an opam switch.

.. note::
   Eventhough the regular kind of context is specified by the ``(default ...)``
   field of the ``context`` stanza, we refer to it as the **regular context** to
   avoid confusion with the default context name.

.. code:: dune

    (context
     (default
      <optional-fields>))

.. code:: dune

    (context
     (opam
      (switch <opam-switch-name>)
      <optional-fields>))

``<optional-fields>`` common to both context types are:

- ``(name <name>)`` is the subdirectory's name for ``_build``, where this
   build's context artifacts will be stored. If omitted, the context name is
   ``default`` for the default context, and switch name for the opam context.

- ``(merlin)`` instructs Dune to use this build context for Merlin.

- ``(generate_merlin_rules)`` instructs Dune to generate Merlin rules for this
  context, even if it is not the one selected via ``(merlin)``.

- ``(profile <profile>)`` sets a different profile for a :term:`build context`.
  This has precedence over the command-line option ``--profile``.

- ``(env <env>)`` sets the environment for a particular context. This is of

>   higher precedence than the root ``env`` stanza in the workspace file. This

Reviewer: This takes precedence over the...

> context

Reviewer: I would suggest the following:

```
 The ``(context ...)`` stanza declares a build context. Each context builds the
  same sources with a different configuration and stores its artefacts in a
  separate subdirectory of ``_build/``. This is useful for cross compilation or
  building against multiple lock directories or opam switches.

  By default, Dune uses a single context called ``default`` with artefacts in
  ``_build/default/``. There are two kinds of context: ``(default ...)`` for a
  regular build context, or ``(opam ...)`` to use an opam switch.
```

This explains somewhat more what a context actually is

Author: This reads better, thanks!

different kinds of context decalarations: ``(default ...)`` for the regular
build context, or ``(opam ...)`` to use an opam switch.

.. note::

>    Eventhough the regular kind of context is specified by the ``(default ...)``

Reviewer 3: ```suggestion
   Despite the regular kind of context is specified by the ``(default ...)``
```
maybe?

Reviewer: I don't think that quite works. Something like:

```
Despite its name, the default field specifies what we call the regular context, named this way to avoid confusion with the default context name.
```

or something similar.

  context.

- ``(host <host_context>)`` chooses a different context to build binaries that
  are meant to be executed on the host machine, such as preprocessors. See
  :ref:`cross-compilation` for more information. This is mutually exclusive with
  the ``(targets ...)`` field.

- ``(targets <targets>)`` lets you specify a list of target names to set up
  cross compilation targets. See :ref:`cross-compilation` for more information.

>   This is mutually exclusive with the ``(host ...)`` field.

Reviewer: specifies target names for cross-compilation

  execution counters is *src/fdo.exe.fdo-profile*.  This feature is
  **experimental** and no backwards compatibility is implied.

``<optional-fields>`` specific to ``(context default)`` are:

-  ``(lock_dir <path>)`` specifies the lock directory that will be used for
   building this context (if any). If no lock directory is specified
   ``dune.lock`` will be used. See the
   :doc:`/reference/dune-workspace/lock_dir` stanza for lock directory
   configuration options.

``<optional-fields>`` specific to ``(context (opam ...))`` are:

-  ``(root <opam-root>)`` is the opam root. By default, it will take the opam
   root defined by the environment in which ``dune`` is run, which is usually
   ``~/.opam``.

> 

Author: Thanks for the review, @Alizter! I've added a point about `instruments_with`. I don't know more information than this, so if there's anything more to be added, let me know.

  execution counters is *src/fdo.exe.fdo-profile*.  This feature is
  **experimental** and no backwards compatibility is implied.

- ``(instrument_with <instrumentation_backend>)`` turns on instrumentation for
  the context. See :doc:`/instrumentation` for more information.

``<optional-fields>`` specific to ``(context default)`` are:

> -  ``(lock_dir <path>)`` specifies the lock directory that will be used for

Reviewer: single space after bullet

  execution counters is *src/fdo.exe.fdo-profile*.  This feature is
  **experimental** and no backwards compatibility is implied.

- ``(instrument_with <instrumentation_backend>)`` turns on instrumentation for
  the context. See :doc:`/instrumentation` for more information.

> - ``(disable_dynamically_linked_foreign_archives <bool>)``: Disable Dune's

Reviewer: ```suggestion
- ``(disable_dynamically_linked_foreign_archives <bool>)`` disables Dune's
```

>   default behavior of building/installing dynamically-linked foreign archives

Reviewer: ```suggestion
  default behavior of building and installing dynamically-linked foreign archives
```
